File: examples/java/src/main/java/com/tokenizers/Encoding.java
package com.tokenizers;

/**
 * Represents an encoded sequence from the tokenizer
 */
public class Encoding implements AutoCloseable {
    
    private long handle;

    public Encoding(long handle) {
        this.handle = handle;
    }

    /**
     * Get the number of tokens
     * @return The length of the encoding
     */
    public int getLength() {
        return nativeGetLength(handle);
    }

    /**
     * Get the token IDs
     * @return Array of token IDs
     */
    public int[] getIds() {
        return nativeGetIds(handle);
    }

    /**
     * Get the tokens as strings
     * @return Array of token strings
     */
    public String[] getTokens() {
        return nativeGetTokens(handle);
    }

    /**
     * Get the type IDs
     * @return Array of type IDs
     */
    public int[] getTypeIds() {
        return nativeGetTypeIds(handle);
    }

    /**
     * Get the special tokens mask
     * @return Array indicating which tokens are special
     */
    public int[] getSpecialTokensMask() {
        return nativeGetSpecialTokensMask(handle);
    }

    /**
     * Get the attention mask
     * @return Array for attention mask
     */
    public int[] getAttentionMask() {
        return nativeGetAttentionMask(handle);
    }

    /**
     * Free the encoding resources
     */
    @Override
    public void close() {
        if (handle != 0) {
            nativeFree(handle);
            handle = 0;
        }
    }

    @Override
    protected void finalize() throws Throwable {
        close();
        super.finalize();
    }

    // Native method declarations
    private static native int nativeGetLength(long handle);
    private static native int[] nativeGetIds(long handle);
    private static native String[] nativeGetTokens(long handle);
    private static native int[] nativeGetTypeIds(long handle);
    private static native int[] nativeGetSpecialTokensMask(long handle);
    private static native int[] nativeGetAttentionMask(long handle);
    private static native void nativeFree(long handle);
}

File: examples/kotlin/src/main/kotlin/com/tokenizers/Encoding.kt
package com.tokenizers

class Encoding internal constructor(private val handle: Long) : AutoCloseable {
    
    companion object {
        @JvmStatic
        private external fun nativeGetLength(handle: Long): Int
        
        @JvmStatic
        private external fun nativeGetIds(handle: Long): IntArray?
        
        @JvmStatic
        private external fun nativeGetTokens(handle: Long): Array<String>?
        
        @JvmStatic
        private external fun nativeGetTypeIds(handle: Long): IntArray?
        
        @JvmStatic
        private external fun nativeGetSpecialTokensMask(handle: Long): IntArray?
        
        @JvmStatic
        private external fun nativeGetAttentionMask(handle: Long): IntArray?
        
        @JvmStatic
        private external fun nativeFree(handle: Long)
    }
    
    val length: Int
        get() = nativeGetLength(handle)
    
    fun getIds(): IntArray {
        return nativeGetIds(handle) ?: throw RuntimeException("Failed to get token IDs")
    }
    
    fun getTokens(): Array<String> {
        return nativeGetTokens(handle) ?: throw RuntimeException("Failed to get tokens")
    }
    
    fun getTypeIds(): IntArray {
        return nativeGetTypeIds(handle) ?: throw RuntimeException("Failed to get type IDs")
    }
    
    fun getSpecialTokensMask(): IntArray {
        return nativeGetSpecialTokensMask(handle) ?: throw RuntimeException("Failed to get special tokens mask")
    }
    
    fun getAttentionMask(): IntArray {
        return nativeGetAttentionMask(handle) ?: throw RuntimeException("Failed to get attention mask")
    }
    
    override fun close() {
        if (handle != 0L) {
            nativeFree(handle)
        }
    }
}